File: temoto_nl_interface/actions/find_coordinates/src/getcoord_py/getcoord_grid_generation.py
#getcoord_grid_generation.py

import numpy as np
import cv2

def get2(cost_map, resolution, grid_scale):
    
    # Calculate pixels per grid cell
    pixels_per_grid_cell = grid_scale  # Assuming grid_scale is in pixels

    # Map dimensions
    height, width = cost_map.shape[:2]

    # Create a copy of the map to draw the grid on
    grid_map = cost_map.copy()

    # Ensure cost_map is in grayscale
    if len(cost_map.shape) == 3 and cost_map.shape[2] == 3:
        cost_map_gray = cv2.cvtColor(cost_map, cv2.COLOR_BGR2GRAY)
    else:
        cost_map_gray = cost_map.copy()

    # Threshold to binary if necessary (assuming white is 255 and black is 0)
    _, binary_map = cv2.threshold(cost_map_gray, 254, 255, cv2.THRESH_BINARY)

    # Define grid line color
    grid_color = (0, 0, 255) 

    # Generate x positions for vertical grid lines
    x_positions = np.arange(0, width, pixels_per_grid_cell).astype(int)

    # Draw vertical grid lines only on white areas
    for x in x_positions:
        # Iterate over each y to set the pixel if it's white
        for y in range(height):
            #if binary_map[y, x] == 255:
            grid_map[y, x] = grid_color

    # Generate y positions for horizontal grid lines
    y_positions = np.arange(0, height, pixels_per_grid_cell).astype(int)

    # Draw horizontal grid lines only on white areas
    for y in y_positions:
        # Iterate over each x to set the pixel if it's white
        for x in range(width):
            #if binary_map[y, x] == 255:
            grid_map[y, x] = grid_color

    return grid_map





def get(cost_map, resolution, grid_scale):
    
    # Calculate pixels per grid cell
    pixels_per_grid_cell = grid_scale  # Assuming grid_scale is in pixels

    # Map dimensions
    height, width = cost_map.shape[:2]

    # Create a copy of the map to draw the grid on
    grid_map = cost_map.copy()

    # Ensure cost_map is in grayscale
    if len(cost_map.shape) == 3 and cost_map.shape[2] == 3:
        cost_map_gray = cv2.cvtColor(cost_map, cv2.COLOR_BGR2GRAY)
    else:
        cost_map_gray = cost_map.copy()

    # Threshold to binary if necessary (assuming white is 255 and black is 0)
    _, binary_map = cv2.threshold(cost_map_gray, 254, 255, cv2.THRESH_BINARY)

    # Define grid line color
    grid_color = (0, 0, 255)  # Gray color

    # Generate x positions for vertical grid lines
    x_positions = np.arange(0, width, pixels_per_grid_cell).astype(int)

    # Draw vertical grid lines only on white areas
    for x in x_positions:
        # Iterate over each y to set the pixel if it's white
        for y in range(height):
            if binary_map[y, x] == 255:
                grid_map[y, x] = grid_color

    # Generate y positions for horizontal grid lines
    y_positions = np.arange(0, height, pixels_per_grid_cell).astype(int)

    # Draw horizontal grid lines only on white areas
    for y in y_positions:
        # Iterate over each x to set the pixel if it's white
        for x in range(width):
            if binary_map[y, x] == 255:
                grid_map[y, x] = grid_color

    return grid_map

File: temoto_nl_interface/actions/find_coordinates/src/get_coordinates_scripts/getcoord_grid_generation.cpp
#include <opencv2/opencv.hpp>
#include <vector>

cv::Mat get2(const cv::Mat& cost_map, int resolution, int grid_scale) {
    int pixels_per_grid_cell = grid_scale;
    int height = cost_map.rows;
    int width = cost_map.cols;

    cv::Mat grid_map = cost_map.clone();

    cv::Mat cost_map_gray;
    if (cost_map.channels() == 3) {
        cv::cvtColor(cost_map, cost_map_gray, cv::COLOR_BGR2GRAY);
    } else {
        cost_map_gray = cost_map.clone();
    }

    cv::Mat binary_map;
    cv::threshold(cost_map_gray, binary_map, 254, 255, cv::THRESH_BINARY);

    cv::Scalar grid_color(0, 0, 255); 

    for (int x = 0; x < width; x += pixels_per_grid_cell) {
        for (int y = 0; y < height; ++y) {
            grid_map.at<cv::Vec3b>(y, x) = cv::Vec3b(0, 0, 255);
        }
    }

    for (int y = 0; y < height; y += pixels_per_grid_cell) {
        for (int x = 0; x < width; ++x) {
            grid_map.at<cv::Vec3b>(y, x) = cv::Vec3b(0, 0, 255);
        }
    }

    return grid_map;
}

cv::Mat get(const cv::Mat& cost_map, int resolution, int grid_scale) {
    int pixels_per_grid_cell = grid_scale;
    int height = cost_map.rows;
    int width = cost_map.cols;

    cv::Mat grid_map = cost_map.clone();

    cv::Mat cost_map_gray;
    if (cost_map.channels() == 3) {
        cv::cvtColor(cost_map, cost_map_gray, cv::COLOR_BGR2GRAY);
    } else {
        cost_map_gray = cost_map.clone();
    }

    cv::Mat binary_map;
    cv::threshold(cost_map_gray, binary_map, 254, 255, cv::THRESH_BINARY);

    for (int x = 0; x < width; x += pixels_per_grid_cell) {
        for (int y = 0; y < height; ++y) {
            if (binary_map.at<uchar>(y, x) == 255) {
                grid_map.at<cv::Vec3b>(y, x) = cv::Vec3b(0, 0, 255);
            }
        }
    }

    for (int y = 0; y < height; y += pixels_per_grid_cell) {
        for (int x = 0; x < width; ++x) {
            if (binary_map.at<uchar>(y, x) == 255) {
                grid_map.at<cv::Vec3b>(y, x) = cv::Vec3b(0, 0, 255);
            }
        }
    }

    return grid_map;
}
